Use SpillSequences when appropriate instead of Sequences

333fred opened the pull request:
Fixes https://github.com/dotnet/roslyn/issues/54704. This means that, instead of expressions that need to `&& true` in the case of builders with void returns but a bool constructor parameter, we now generate `if (builderConstructorSuccess) { appendCalls }`. As part of this, I simplified the code around how the sequence/spillsequence is actually generated to reduce unnecessary sequence nesting, but that had no affect on the final codegen.

Commit 2 is a quick fix for part of https://github.com/dotnet/roslyn/issues/56044.

333fred commented:
@chsienki @dotnet/roslyn-compiler for review, please.

tmat commented on src/Compilers/Core/Portable/SynthesizedLocalKind.cs:
```diff
@@ -28,6 +29,12 @@ namespace Microsoft.CodeAnalysis
     /// </remarks>
     internal enum SynthesizedLocalKind
     {
+        /// <summary>
```
Can we just use `LoweringTemp` instead of adding a new kind? Usually, we don't need to distinguish between different kinds of short-lived temps introduced during lowering.

333fred replied:
No, we cannot. These can be ref structs and used in async methods.

tmat replied:
It's not clear from the description of LoweringTemp why would that be an issue.

AlekseyTs replied:
> No, we cannot. These can be ref structs and used in async methods. 

I do not see this enum member used in any meaningful way. We attach it to the local symbol, but never check for this value.

tmat commented on src/Compilers/Core/Portable/SynthesizedLocalKind.cs:
```diff
@@ -14,7 +14,8 @@ namespace Microsoft.CodeAnalysis
     ///    The lifespan of a temporary variable shall not cross a statement boundary (a PDB sequence point).
-    ///    These variables are not tracked by EnC and don't have names.
+    ///    These variables are not tracked by EnC and don't have names. Only values less than 0 are considered
+    ///    short-lived: new short-lived kinds should have a negative value.
```
It shouldn't be necessary to introduce new short-lived kinds.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs:
```diff
@@ -130,44 +130,30 @@ private BoundExpression RewriteInterpolatedStringConversion(BoundConversion conv
 
-                resultExpressions.Add(handlerConstructionAssignment);
+                // if (appendShouldProceedLocal) { appendCallsStatements }
+                var resultIf = _factory.If(appendShouldProceedLocal, _factory.Block(appendCallsStatements));
```
Can this be just a statement list?

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs:
```diff
@@ -130,44 +130,30 @@ private BoundExpression RewriteInterpolatedStringConversion(BoundConversion conv
-                // Odd case of no append calls, but with an out param. We don't need to generate any jumps checking the local because there's
-                // nothing to short circuit and avoid, but we do need a sequence to hold the lifetime of the local
-                resultExpressions.Add(_factory.Sequence(ImmutableArray.Create(appendShouldProceedLocal.LocalSymbol), ImmutableArray<BoundExpression>.Empty, handlerConstructionAssignment));
+                _needsSpilling = true;
```
>_needsSpilling = true; [](http://example.com/codeflow?start=16&length=22)

I think this should be set where we create the spill sequence.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs:
```diff
@@ -130,44 +130,30 @@ private BoundExpression RewriteInterpolatedStringConversion(BoundConversion conv
-                // nothing to short circuit and avoid, but we do need a sequence to hold the lifetime of the local
-                resultExpressions.Add(_factory.Sequence(ImmutableArray.Create(appendShouldProceedLocal.LocalSymbol), ImmutableArray<BoundExpression>.Empty, handlerConstructionAssignment));
+                _needsSpilling = true;
+                return new InterpolationHandlerResult(ImmutableArray.Create(_factory.ExpressionStatement(handlerConstructionAssignment), resultIf), builderTemp, appendShouldProceedLocal.LocalSymbol, _factory);
```
>ImmutableArray.Create [](http://example.com/codeflow?start=54&length=21)

Do we ever pass more than one statement? Consider avoiding array allocation otherwise.

AlekseyTs commented:
Done with review pass (commit 2)

333fred commented:
@AlekseyTs @tmat addressed feedback, PTAL.

---
In reply to: [912718331](https://github.com/dotnet/roslyn/pull/56153#issuecomment-912718331) [](http://example.com/codeflow?ancestors=912718331,912718331)

AlekseyTs reviewed (approved):
LGTM (commit 4)

333fred commented:
@chsienki @dotnet/roslyn-compiler for a second review please.

333fred commented:
@dotnet/roslyn-compiler @chsienki please review.

333fred commented:
/azp run
